src/compiler/checker.ts

```diff
@@ -36437,6 +36443,9 @@ namespace ts {
                 return type;
             }
             if (isGenericObjectType(objectType)) {
+                if (isTypeAssignableTo(indexType, getImmediateIndexType(apparentObjectType, /*stringsOnly*/ false))) {

```

Comment:
I don't fully expect this to be the final solution - but I think that some form of this is sound. I'm just unsure if I should avoid doing this in some situations or if there is a better way to compare the available keys on this `apparentObjecType`/`objectType` with the `indexType`.